db/hints/manager.cc

```diff
@@ -569,4 +576,85 @@ future<> manager::with_file_update_mutex_for(endpoint_id ep, noncopyable_functio
+    // It's imperative because we cannot allow for a host ID to have mulitple IPs while
+    // this function is being executed.
+    //
+    // Imagine the following scenario of consecutive steps to visualize it:
+    //  (1) We rename the directory corresponding to ip_1 to host_id_1.
+    //  (2) The node corresponding to host_id_1 changes its IP to ip_2.
+    //  (3) We move on to processing a directory correspondiong to ip_2.
+    //      At this step, a failure occurs because ::rename(2) (so `seastar::rename_file` by extension too)
+    //      assumes that the new directory name (or more specifically, its path) doesn't exist.
+    //
+    // We utilize the useful property of token metadata here that ensures the topology stays immutable
+    // as long as we access it via this pointer.
+    auto tmptr = _proxy.get_token_metadata_ptr();
+
+    try {
+        // When upgrading Scylla, rename directories so that all of them use host IDs.
+        // If there is no mapping from an IP to a host ID, drop a directory by effectively removing it
+        // and its contents. Hinted Handoff is a "best-effort" mechanism -- there is no reason to deal
+        // with such edge cases, especially because this can only happen once -- during an upgrade.
+        co_await lister::scan_dir(_hints_dir, lister::dir_entry_types::of<directory_entry_type::directory>(),
+                [tmptr] (fs::path datadir, directory_entry de) -> future<> {
+            auto remove_directory = [] (fs::path p, const std::string_view reason) -> future<> {
+                // We cannot map the IP to the corresponding host ID either because
+                // the relevant mapping doesn't exist anymore or an error occurred. Drop it.
+                //
+                // We only care about directories named after IPs during an upgrade.
+                // In normal circumstances, this whole lambda should be equivalent to a no-op,
+                // i.e. it should have no side effects.
+                manager_logger.warn("Removing hint directory {}. Reason: {}", p.filename(), reason);
+                return lister::rmdir(p);
+            };
+
+            auto parse_name = [] (const sstring& name) -> std::optional<locator::host_id_or_endpoint> {
+                try {
+                    return locator::host_id_or_endpoint{name};
+                } catch (...) {
+                    return {};
+                }
+            };
+
+            // Old path: `<...>/<shard id>/<IP>`.
+            const fs::path current_path = datadir / de.name;
+            auto maybe_parsed_name = parse_name(de.name);
+
+            if (!maybe_parsed_name) {
+                return remove_directory(current_path,
+                        seastar::format("The directory's name {} is neither IP, nor host ID.", de.name));
+            }
+            
+            if (maybe_parsed_name.value().has_host_id()) {
+                return make_ready_future<>();
+            }
```

Comment:
I'm not sure how `scan_dir/getdents64` behaves when we mutate the files it's iterating over (renaming/removing). I couldn't find this in the man-s.

I think it would be conceptually simpler to do the migration in phases:
1. readonly run of `scan_dir`, fill `files_to_rename/files_to_remove` vectors;
2. run all renames, append all the failures to the `files_to_remove`;
3. run all removes;
4. run one single `co_await io_check(sync_directory, _hints_dir.native());` when we are done.

Reply:
Sensible. It's also something that is liable to depend on the filesystem and can change with implementations.